### PR TITLE
feat: support reading partitioned Delta table.

### PR DESCRIPTION
**Pull request by youngsofun**
I hereby agree to the terms of the CLA available at: https://docs.databend.com/dev/policies/cla/

## Summary

In a delta table, partition columns are not stored in parquet file.
so it needs a few efforts to make pushdown work:
- context:
  - Table store partition column names in meta.engine_options.
  - Each partition carries all partition column values in the same order.
  - With this order, we can get needed info with a PartitionIndex.
- pushdown:
  - projections (mask): partition columns are excluded when read parquet file and inserted at last. (use FieldIndex->PartitionIndex)
  - filter pass to parquet reader: all partition columns are appended to the filter input columns. （ID=PartitionIndex+ num_of_non_partition_input_columns）
  - pruner: ColumnRef of partition columns in filter expr are replace with const scalars (by name).

Type of partition columns can only be simple primitive types.

## Tests

- [x] Unit Test
- [x] Logic Test
- [ ] Benchmark Test
- [ ] No Test  - _Explain why_

## Type of change

- [ ] Bug Fix (non-breaking change which fixes an issue)
- [x] New Feature (non-breaking change which adds functionality)
- [ ] Breaking Change (fix or feature that could cause existing functionality not to work as expected)
- [ ] Documentation Update
- [ ] Refactoring
- [ ] Performance Improvement
- [ ] Other (please describe):

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/14084)

### CODE REVIEW

File: src/query/expression/src/expression.rs [hunk 1]
```diff
@@ -361,6 +361,88 @@ impl<Index: ColumnIndex> Expr<Index> {
         }
     }
 
+    pub fn fill_const_column(&self, consts: &HashMap<Index, Scalar>) -> Expr<Index> {
```
[the remark this answers is not in the document]
Comment (RinChanNOWWW) [reply src/query/expression/src/expression.rs:364]: Yes, we can construct a constant domain and then use constant folder to fold the expr.

[the remark this answers is not in the document]
Comment (youngsofun) [reply src/query/expression/src/expression.rs:364]: did not found a util to convert const scalar to constant domain, `fill_const_column` seemed a litte eaiser for less match branches.

[the remark this answers is not in the document]
Comment (sundy-li) [reply src/query/expression/src/expression.rs:364]: You can refer to `HivePartitionPruner::get_column_stats` in prune.

[the remark this answers is not in the document]
Comment (youngsofun) [reply src/query/expression/src/expression.rs:364]: if you prefer constant folder， I can add fn ` Domain::from_scalar`

[the remark this answers is not in the document]
Comment (sundy-li) [reply src/query/expression/src/expression.rs:364]: We have `ColumnStatistics` --> `Domain`,  so we just need to add `scalar` --> `ColumnStatistics` . This is already covered in `HivePartitionPruner::get_column_stats`.

`fill_const_column` only injects the value into the expr, you still need  `constant folder` even with `fill_const_column`.  

For example:

`p1 > p2 or p2 < 1`

[the remark this answers is not in the document]
Comment (sundy-li) [reply src/query/expression/src/expression.rs:364]: Let's keep this.

**Review by sundy-li — APPROVED** [review]

**Pull request merged**